homeassistant/components/flume/coordinator.py

```diff
@@ -33,3 +41,53 @@ async def _async_update_data(self) -> None:
             self.flume_device.values,
             self.flume_device.query_payload,
         )
+
+
+class FlumeNotificationDataUpdateCoordinator(DataUpdateCoordinator[None]):
+    """Data update coordinator for flume notifications."""
+
+    def __init__(self, hass: HomeAssistant, auth) -> None:
+        """Initialize the Coordinator."""
+        super().__init__(
+            hass,
+            name=DOMAIN,
+            logger=_LOGGER,
+            update_interval=NOTIFICATION_SCAN_INTERVAL,
+        )
+        self.auth = auth
+        self.notifications: dict = {}
+        self.active_notifications_by_device: dict = {}
+
+    def _update_lists(self):
+        """Query flume for notification list."""
+        self.notifications = pyflume.FlumeNotificationList(
+            self.auth, read="true"
+        ).notification_list
+        _LOGGER.debug("Notifications %s", self.notifications)
+
+        # The list of notifications returned by API will be in chronological order and there may be
+        # multiple notifications of the same type (such as High Flow). The only special
+        # case here is bridge notifications: NOTIFICATION_BRIDGE_DISCONNECT, which will be either True
+        # or False as indicated in the ["extra"][BRIDGE_NOTIFICATION_KEY] field.
+
+        notifications_by_device = {}
+        for item in self.notifications:
+
+            device_id = item["device_id"]
+            rule = item["extra"]["event_rule_name"]
+
+            # Bridge notifications are a special case, the actual status (connected/disconnected) is stored
+            # in the BRIDGE_NOTIFICATION_KEY and we get a notification
+            value = item["extra"].get(BRIDGE_NOTIFICATION_KEY, True)
+            notifications_by_device.setdefault(device_id, {})
```

Comment:
```suggestion
            notifications_by_device.setdefault(device_id, set())
```

I was thinking you could make this a `set()` and then only add the ones to the set that are in the `on` state.

Then your binary sensor check becomes a simple check to see if its in the set or not

Reply:
So the main issue I'm running into is I need to process ALL notifications first before I can turn them into a set. I did, however, make updates that stores a set of active notifications

Bridge notifications are basically:  "Is it disconnected", so a True means disconnected but if it reconnects you end up with a False notification - so after processing through that (via a dictionary) - I can then capture a set of active notifications (which I've done).

Reply:
Here is an example of how the bridge stuff comes in:


```json
{
    "created_datetime": "2022-07-17T23:20:48.000Z",
    "device_id": "xxxxxxxxxxxx",
    "extra": {
      "connected": False,
      "event_rule_name": "Bridge Disconnection"
    },
    "id": 8267532,
    "message": "Your Flume device will not be able to monitor your water usage or detect leaks until you reconnect your Bridge. To reconnect your device, open your Flume App and follow the instructions on your Dashboard.",
    "read": True,
    "title": "Your Bridge at Home has Lost WiFi Connection!",
    "type": 8,
    "user_id": xxxx
  },
{
    "created_datetime": "2022-07-18T05:22:30.000Z",
    "device_id": "xxxxxxxxxxxx",
    "extra": {
      "connected": True,
      "event_rule_name": "Bridge Disconnection"
    },
    "id": 8272675,
    "message": "Your Water Sensor should reconnect to your Bridge on its own in approximately 40 minutes. Within an hour of your Water Sensor reconnecting, your app will begin backfilling water usage!",
    "read": True,
    "title": "Your Bridge at Home has reconnected to WiFi!",
    "type": 8,
    "user_id": xxxx
  },
```

Reply:
```diff
diff --git a/homeassistant/components/flume/binary_sensor.py b/homeassistant/components/flume/binary_sensor.py
index e72901df39..f8c9c2ebd3 100644
--- a/homeassistant/components/flume/binary_sensor.py
+++ b/homeassistant/components/flume/binary_sensor.py
@@ -116,19 +116,8 @@ class FlumeBinarySensor(FlumeEntity, BinarySensorEntity):
     @property
     def is_on(self) -> bool:
         """Return on state."""
-
-        rule = self.entity_description.event_rule
-
-        # The default notification value for notifications will be False, which means the notification
-        # is not currently active. Bridge notifications, on the other hand, which are of Connectivity
-        # class need to default to true which shows the unit IS connected. In the absence of an actual
-        # notification from the coordinator this value will be used - as the "initial" or default value
-        initial_value = (
-            self.entity_description.device_class == BinarySensorDeviceClass.CONNECTIVITY
-        )
-
-        return (
-            rule
-            in self.coordinator.active_notifications_by_device.get(self.device_id, [])
-            or initial_value
-        )
+        if notifications := self.coordinator.active_notifications_by_device.get(
+            self.device_id
+        ):
+            return False
+        return self.entity_description.event_rule in notifications
diff --git a/homeassistant/components/flume/coordinator.py b/homeassistant/components/flume/coordinator.py
index f3ab8d13d5..85c8aec2ed 100644
--- a/homeassistant/components/flume/coordinator.py
+++ b/homeassistant/components/flume/coordinator.py
@@ -69,7 +69,7 @@ class FlumeNotificationDataUpdateCoordinator(DataUpdateCoordinator[None]):
         # multiple notifications of the same type (such as High Flow). The only special
         # case here is bridge notifications: NOTIFICATION_BRIDGE_DISCONNECT, which will be either True
         # or False as indicated in the ["extra"][BRIDGE_NOTIFICATION_KEY].
-        full_notification_info = {}
+        notifications_by_device: dict[str, set[str]] = {}
         for item in self.notifications:
 
             device_id = item["device_id"]
@@ -78,21 +78,10 @@ class FlumeNotificationDataUpdateCoordinator(DataUpdateCoordinator[None]):
             # Because bridge is a disconnect notification, True = disconnected - its value needs to be reversed
             # all other notifications will be not False ... aka True
             value = not item["extra"].get(BRIDGE_NOTIFICATION_KEY, False)
-
-            full_notification_info.setdefault(device_id, {})
-            full_notification_info[device_id][rule] = value
+            if value:
+                notifications_by_device.setdefault(device_id, set()).add(rule)
 
         # Next preserve only notifications that are present in a set
-        notifications_by_device = {}
-        for device_id in full_notification_info.items():
-            notifications_by_device[device_id] = set(
-                dict(
-                    filter(
-                        lambda x: x[1] is True,
-                        full_notification_info[device_id].items(),
-                    )
-                )
-            )
 
         self.active_notifications_by_device = notifications_by_device
 
```

I think you can make this much simpler with something like the above (you may need to adjust it for the bridge notification key)